File: host/Preview/FlipBudget.cs
using System.Windows;
using Eiviz.Host.I18n;

namespace Eiviz.Host.Preview;

internal static class FlipBudget
{
    public const int AutoDefault = 8;
    private const long LearnWindowMs = 3000;

    private static uint _limitSetting;
    private static int _ceiling = AutoDefault;
    private static int _attached;
    private static SwapchainHost? _lastAttach;
    private static long _attachTick;
    private static ulong _seenLost;
    private static bool _refusedOnce;

    public static void Configure(uint limit)
    {
        _limitSetting = IsAllowed(limit) ? limit : 0u;
        _ceiling = _limitSetting == 0
            ? GpuPresentStore.ObservedCeiling ?? AutoDefault
            : (int)_limitSetting;
    }

    public static bool TryBegin(SwapchainHost host)
    {
        if (!OperatingSystem.IsWindows())
            return true;
        if (_attached >= EffectiveMax())
        {
            ShowRefuse();
            return false;
        }
        _attached++;
        _lastAttach = host;
        _attachTick = Environment.TickCount64;
        return true;
    }

    public static void Cancel(SwapchainHost host)
    {
        if (!OperatingSystem.IsWindows())
            return;
        if (_attached > 0)
            _attached--;
        if (ReferenceEquals(_lastAttach, host))
            _lastAttach = null;
    }

    public static void End(SwapchainHost host)
    {
        Cancel(host);
    }

    public static void ObserveLost(ulong total)
    {
        if (!OperatingSystem.IsWindows() || total <= _seenLost)
            return;
        _seenLost = total;
        if (_lastAttach is null)
            return;
        if (_attached <= 2)
            return;
        if (Environment.TickCount64 - _attachTick > LearnWindowMs)
            return;
        var victim = _lastAttach;
        victim.ReleaseNative();
        if (_limitSetting == 0 && _attached >= 2)
            GpuPresentStore.Save(_attached);
        _ceiling = Math.Max(2, _attached);
    }

    private static int EffectiveMax() =>
        _limitSetting == 0 ? _ceiling : (int)_limitSetting;

    private static bool IsAllowed(uint limit) =>
        limit is 0 or 4 or 6 or 8 or 10 or 12 or 16;

    private static void ShowRefuse()
    {
        if (_refusedOnce)
            return;
        _refusedOnce = true;
        var app = Application.Current;
        if (app is null)
            return;
        app.Dispatcher.BeginInvoke(() =>
        {
            var owner = app.MainWindow;
            if (owner is null)
                MessageBox.Show(Loc.T("msg.flipBudget"));
            else
                MessageBox.Show(owner, Loc.T("msg.flipBudget"));
        });
    }
}

File: mac/Sources/EivizMac/FlipBudget.swift
import AppKit
import Foundation

enum GpuPresentStore {
    private struct Dto: Codable {
        var observedCeiling: Int
    }

    nonisolated(unsafe) private(set) static var observedCeiling: Int?

    static func load() {
        let url = storeURL()
        guard FileManager.default.fileExists(atPath: url.path),
              let data = try? Data(contentsOf: url),
              let dto = try? JSONDecoder().decode(Dto.self, from: data),
              (2...16).contains(dto.observedCeiling)
        else { return }
        observedCeiling = dto.observedCeiling
    }

    static func save(_ ceiling: Int) {
        guard ceiling >= 2 else { return }
        observedCeiling = ceiling
        let url = storeURL()
        do {
            try FileManager.default.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
            let data = try JSONEncoder().encode(Dto(observedCeiling: ceiling))
            try data.write(to: url)
        } catch {
            HostLog.write("WARN", "gpu-present save failed: \(error.localizedDescription)")
        }
    }

    private static func storeURL() -> URL {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("eiviz", isDirectory: true)
            .appendingPathComponent("gpu-present.json")
    }
}

enum FlipBudget {
    static let autoDefault = 6
    private static let learnWindowMs: UInt64 = 3000

    nonisolated(unsafe) private static var limitSetting: UInt32 = 0
    nonisolated(unsafe) private static var ceiling = autoDefault
    nonisolated(unsafe) private static var attached = 0
    nonisolated(unsafe) private static weak var lastAttach: MetalSurfaceView?
    nonisolated(unsafe) private static var attachTick: UInt64 = 0
    nonisolated(unsafe) private static var seenLost: UInt64 = 0

    static func configure(_ limit: UInt32) {
        limitSetting = isAllowed(limit) ? limit : 0
        ceiling = limitSetting == 0
            ? min(GpuPresentStore.observedCeiling ?? autoDefault, autoDefault)
            : Int(limitSetting)
    }

    static func tryOpen(_ surfaces: Int) -> Bool {
        guard surfaces > 0 else { return true }
        if attached + surfaces <= effectiveMax() {
            return true
        }
        showRefuse()
        return false
    }

    static func tryBegin(_ host: MetalSurfaceView) -> Bool {
        if attached >= effectiveMax() {
            return false
        }
        attached += 1
        lastAttach = host
        attachTick = nowMs()
        return true
    }

    static func cancel(_ host: MetalSurfaceView) {
        if attached > 0 {
            attached -= 1
        }
        if lastAttach === host {
            lastAttach = nil
        }
    }

    static func end(_ host: MetalSurfaceView) {
        cancel(host)
    }

    static func observeLost(_ total: UInt64) {
        guard total > seenLost else { return }
        seenLost = total
        guard let victim = lastAttach, attached > 2 else { return }
        guard nowMs() &- attachTick <= learnWindowMs else { return }
        victim.releaseNative()
        if limitSetting == 0, attached >= 2 {
            GpuPresentStore.save(attached)
        }
        ceiling = max(2, attached)
    }

    private static func effectiveMax() -> Int {
        limitSetting == 0 ? ceiling : Int(limitSetting)
    }

    private static func isAllowed(_ limit: UInt32) -> Bool {
        limit == 0 || limit == 4 || limit == 6 || limit == 8 || limit == 10 || limit == 12 || limit == 16
    }

    private static func showRefuse() {
        let alert = NSAlert()
        alert.messageText = L10n.t("msg.flipBudget")
        alert.alertStyle = .informational
        alert.runModal()
    }

    private static func nowMs() -> UInt64 {
        DispatchTime.now().uptimeNanoseconds / 1_000_000
    }
}
